mplc/multi_partner_learning.py

```diff
@@ -5,804 +5,478 @@
+    def build_model_from_weights(self, new_weights):
+        """Generate a new model initialized with weights passed as arguments"""
+        new_model = self.generate_new_model()
+        new_model.set_weights(new_weights)
+        return new_model
 
-        # Train model (iterate for each epoch and mini-batch)
-        for epoch_index in range(epoch_count):
+    def init_model(self):
+        new_model = self.generate_new_model()
 
-            self.epoch_index = epoch_index
+        if self.use_saved_weights:
+            logger.info("Init model with previous coalition model")
+            new_model.load_weights(self.init_model_from)
+        else:
+            logger.info("Init new model")
 
-            # Clear Keras' old models (except if the approach is sequential and the model has to persist across epochs)
-            if self.learning_approach not in ['seq-pure', 'seq-with-final-agg']:
-                clear_session()
+        return new_model
 
-            # Split the train dataset in mini-batches
-            self.split_in_minibatches()
+    def early_stop(self):
+        logger.debug("      Checking if early stopping criteria are met:")
+        if self.is_early_stopping:
+            # Early stopping parameters
+            if (
+                    self.epoch_index >= constants.PATIENCE
+                    and self.history.history['mpl_model']['val_loss'][-1, -1] >
+                    self.history.history['mpl_model']['val_loss'][-constants.PATIENCE, -1]
+            ):
+                logger.debug("         -> Early stopping criteria are met, stopping here.")
+                return True
+            else:
+                logger.debug("         -> Early stopping criteria are not met, continuing with training.")
+        else:
+            return False
 
-            # Iterate over mini-batches and train
-            for minibatch_index in range(minibatch_count):
+    def fit(self):
+        """Return the score on test data of a final aggregated model trained in a federated way on each partner"""
 
-                self.minibatch_index = minibatch_index
+        start = timer()
+        # Train model (iterate for each epoch and mini-batch)
+        while self.epoch_index < self.epoch_count:
 
-                if self.learning_approach == 'fedavg':
-                    self.compute_collaborative_round_fedavg()
+            self.fit_epoch()  # perform an epoch on the self.model
 
-                elif self.learning_approach in ['seq-pure', 'seq-with-final-agg']:
-                    self.compute_collaborative_round_sequential(sequentially_trained_model)
+            if self.early_stop():
+                break
+            self.epoch_index += 1
 
-                elif self.learning_approach == 'seqavg':
-                    self.compute_collaborative_round_seqavg()
+        # After last epoch or if early stopping was triggered, evaluate model on the global testset
+        self.history.log_final_model_perf()
```

Comment:
cf previous comment, we should evaluate the model with self.evaluate() then log the perf.